Created workflows best practices (#13991)
PostHog/posthog.com: contents/docs/workflows/best-practices.mdx

Before sending emails or other messages, [set up and verify your channels](/docs/workflows/configure-channels). For email: configure "from" name, from-address, and DNS (SPF/DKIM) to avoid spam/folder issues. 

Use templating/personalization carefully (e.g. with placeholders like `{{ person.name }}`), but also verify that placeholders resolve and avoid sending messages with unresolved variables. 

If sending to external channels (e.g. Slack, SMS, webhooks), treat them as equally important as email: ensure you have fallback or error handling logic (e.g. when a webhook endpoint fails).

## 5. Plan for maintainability: naming, versioning, cleanup

Give workflows descriptive names. E.g. `welcome_email_on_signup`, `drip_7_day_followup`, etc. Avoid ambiguous names like `workflow1`. Version your workflows (e.g. when you make significant logic changes), or at least document changes in a changelog, so you know which version ran when.

When a workflow becomes obsolete (e.g. onboarding process changed, or a campaign ended), archive or delete it to avoid accidental triggering and reduce maintenance burden.

## 6. Minimize duplication and side-effects

> Don’t duplicate workflows that do essentially the same thing under slightly different triggers, which  increases risk of duplicate messages, conflicting updates, or unexpected user experience. 

Reviewer: ```suggestion
Don’t duplicate workflows that do essentially the same thing under slightly different triggers, which increases the risk of duplicate messages, conflicting updates, or unexpected user experience. 
```

Before a full rollout: test with internal users or a small percentage of real users. Monitor delivery (for emails/messages), event logs, user property updates, or other side-effects. Watch for unintended behavior: e.g. multiple emails, loops (if your workflow triggers events that retrigger the same workflow), or delays stacking up.

## 8. Monitor and log

Ensure that failures (e.g. email bounce, webhook error) are captured and surfaced. When facing a problem, check the “[Troubleshooting & FAQs](/docs/workflows/troubleshooting)” section for Workflows.

For long-running workflows (with delays), track user state (which step they are in), and provide a way to debug or abort if something goes wrong.
Periodically review metrics: how many messages sent, how many workflows triggered, bounce/failure rate, etc.

<ProductScreenshot
    imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_12_05_at_14_38_35_2x_908da89b7a.png"
    imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_12_05_at_14_38_35_2x_908da89b7a.png"
    alt="Workflow monitoring and metrics"
    classes="rounded"
/>

## 9. Be mindful of costs and rate limits

> PostHog plans to price Workflows based on the number of real-time destinations and monthly messages after beta. Design workflows to minimize unnecessary messages, e.g. avoid sending multiple messages for the same user around the same event, use batching or deduplication logic if possible. 

Reviewer: ```suggestion
PostHog plans to price Workflows based on the number of real-time destinations and monthly messages after beta. Design workflows to minimize unnecessary messages. For example, avoid sending multiple messages to the same user around the same event, and use batching or deduplication logic where possible. 
```
